zmon_agent/discovery/kubernetes/cluster.py

```diff
@@ -1129,3 +1142,76 @@ def get_postgresql_databases(cluster_id, alias, environment, region, infrastruct
                 entities.append(replicatype_entity)
 
     return entities
+
+
+@trace(tags={'kubernetes': 'hpa'}, pass_span=True)
+def get_cluster_hpas(kube_client, cluster_id, alias, environment, region, infrastructure_account,
+                     namespace=None, **kwargs) -> list:
+    current_span = extract_span_from_kwargs(**kwargs)
+    entities = []
+
+    hpas = get_all(kube_client, kube_client.get_hpas, namespace, span=current_span)
+
+    for h in hpas:
+        obj = h.obj
+
+        conditions_annotation = obj['metadata']['annotations']['autoscaling.alpha.kubernetes.io/conditions']
+        conditions_list = json.loads(conditions_annotation)
+        conditions = {c['type']: c['status'] == 'True' for c in conditions_list}
+
+        entity = {
+            'id': 'hpa-{}-{}[{}]'.format(h.name, h.namespace, cluster_id),
+            'type': HPA_TYPE,
+            'kube_cluster': cluster_id,
+            'alias': alias,
+            'environment': environment,
+            'created_by': AGENT_TYPE,
+            'infrastructure_account': infrastructure_account,
+            'region': region,
+
+            'hpa_name': h.name,
+            'hpa_namespace': obj['metadata']['namespace'],
+
+            'hpa_desired_replicas': obj['status']['desiredReplicas'],
+            'hpa_current_replicas': obj['status']['currentReplicas'],
+            'hpa_min_replicas': obj['spec']['minReplicas'],
+            'hpa_max_replicas': obj['spec']['maxReplicas'],
+            'able_to_scale': conditions.get('AbleToScale', False),
+            'scaling_active': conditions.get('ScalingActive', False),
+            'scaling_limited': conditions.get('ScalingLimited', False),
+        }
+
+        entity.update(entity_labels(obj, 'labels', 'annotations'))
+
+        entities.append(entity)
+
+    return entities
+
+
+@trace(tags={'kubernetes': 'credentialset'}, pass_span=True)
+def get_cluster_credential_sets(kube_client, cluster_id, alias, environment, region, infrastructure_account,
+                                namespace=None, **kwargs) -> list:
+    current_span = extract_span_from_kwargs(**kwargs)
+    entities = []
+
+    credential_sets = get_all(kube_client, kube_client.get_platformcredentialsets, namespace, span=current_span)
+    for cs in credential_sets:
+        obj = cs.obj
+        entity = {
+            'id': 'pcs-{}-{}[{}]'.format(cs.name, cs.namespace, cluster_id),
+            'type': CREDENTIALSET_TYPE,
+            'kube_cluster': cluster_id,
+            'alias': alias,
+            'environment': environment,
+            'created_by': AGENT_TYPE,
+            'infrastructure_account': infrastructure_account,
+            'region': region,
+            'pcs_name': cs.name,
+            'pcs_namespace': obj['metadata']['namespace'],
+            'errors': obj['status']['errors'],
+            'problems': obj['status']['problems'],
```

Comment:
are we sure `['status']['errors']` and `['propblems']` are always there? if not entirely sure then `.get()` would make more sense here.

Reply:
`problems` and `errors` are always present in the `status`. It's part of the definition and defaults to empty list.